scripts/generators/typescript.js

```diff
@@ -0,0 +1,186 @@
+"use strict";
+
+const t = require("../../packages/babel-types");
+const utils = require("./utils");
+
+let code = `// NOTE: This file is autogenerated. Do not modify.
+// See scripts/generators/typescript.js for script used.
+
+export interface Comment {
+  value: string;
+  start: number;
+  end: number;
+  loc: SourceLocation;
+  type: "BlockComment" | "LineComment";
+}
+
+export interface BlockComment extends Comment {
+  type: "BlockComment";
+}
+
+export interface LineComment extends Comment {
+  type: "LineComment";
+}
+
+export interface SourceLocation {
+  start: {
+    line: number;
+    column: number;
+  };
+
+  end: {
+    line: number;
+    column: number;
+  };
+}
+
+interface BaseNode {
+  leadingComments: ReadonlyArray<Comment> | null;
+  innerComments: ReadonlyArray<Comment> | null;
+  trailingComments: ReadonlyArray<Comment> | null;
+  start: number | null;
+  end: number | null;
+  loc: SourceLocation | null;
+  type: Node["type"];
```

Comment:
...but it is here. If you have a `BaseNode`, comparing its `type` is completely useless.

Though arguably you're supposed to use the `t.isFoo` functions instead; and this interface is not exported so users can't use it by accident.

Reply:
I'm actually not sure how safe it is in practice to write it like this because of the circular dependency. `BaseNode` depends on `Node` which depends on (list of interfaces all of) which depend on `BaseNode`.

Reply:
It seems to compile fine. `BaseNode` is not exported, but `Node` (which is refinable union of tagged types) is.